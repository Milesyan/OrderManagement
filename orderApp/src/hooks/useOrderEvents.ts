import useSocket from './useSocket';
import { useImmer } from "use-immer";
import {IOrder} from 'src/types';
import {enableMapSet} from 'immer';
import CONFIGS from '../env';
enableMapSet();

export interface IOrderData {
  idOrderMap: Map<string, IOrder>;
  priceIdMap: Map<number, Set<string>>;
}

export const orderDataMutableUpdater = (origin: IOrderData, orders: IOrder[]) => {
  for (const order of orders) {
    // update price to orderId Map
    if (!origin.priceIdMap.get(order.price)) {
      origin.priceIdMap.set(order.price, new Set([order.id]))
    } else {
      origin.priceIdMap.get(order.price)?.add(order.id)
    }
    // update Id to order Map
    origin.idOrderMap.set(order.id, order)
  }
}
export default function useOrderEvents(): IOrderData{
  const [orderData, setOrderData] = useImmer<IOrderData>({
    idOrderMap: new Map<string, IOrder>(),
    priceIdMap: new Map<number, Set<string>>()
  })
  const updateCallback = (orders: IOrder[]) => {
    setOrderData(draft => {
      orderDataMutableUpdater(draft, orders)
    })
  }
  useSocket(CONFIGS.ORDER_SERVER_HOST, [
    {
      eventName: 'order_event', 
      callback: updateCallback
    }
  ])
  return orderData
}
