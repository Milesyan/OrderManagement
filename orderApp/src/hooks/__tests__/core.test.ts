import {orderDataMutableUpdater} from '../useOrderEvents';

test('test orderDataMutableUpdater logic', () => {
  const data = {
    idOrderMap: new Map(),
    priceIdMap: new Map()
  };
  orderDataMutableUpdater(data, [{
    "customer": "Karina Lopez",
    "destination": "89548 Andrew Road Apt. 797, Lake Phillipville, CA 92007",
    "event_name": "COOKED",
    "id": "b11b8a69",
    "item": "Fish tacos",
    "price": 8700,
    "sent_at_second": 30
  },
  {
    "customer": "Heather Hughes",
    "destination": "7465 Willis Fork, Lake Gabrielle, CA 92816",
    "event_name": "CREATED",
    "id": "0b63508d",
    "item": "Berry smoothie",
    "price": 8029,
    "sent_at_second": 5
  }])
  expect(data).toMatchSnapshot();
  // update status of same id
  orderDataMutableUpdater(data, [{
    "customer": "Karina Lopez",
    "destination": "89548 Andrew Road Apt. 797, Lake Phillipville, CA 92007",
    "event_name": 'DRIVER_RECEIVED',
    "id": "b11b8a69",
    "item": "Fish tacos",
    "price": 8700,
    "sent_at_second": 30
  }])
  expect(data).toMatchSnapshot();
  expect(data.idOrderMap.get('b11b8a69')?.event_name).toBe('DRIVER_RECEIVED');
  // Add new order with same price
  orderDataMutableUpdater(data, [{
    "customer": "Karina Lopez",
    "destination": "89548 Andrew Road Apt. 797, Lake Phillipville, CA 92007",
    "event_name": 'CREATED',
    "id": "b11b8a69xxx",
    "item": "Fish tacos",
    "price": 8700,
    "sent_at_second": 30
  }])
  expect(data).toMatchSnapshot();
  expect(data.priceIdMap.get(8700)).toEqual(new Set(['b11b8a69xxx', 'b11b8a69']))
});