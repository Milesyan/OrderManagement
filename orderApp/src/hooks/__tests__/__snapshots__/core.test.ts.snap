// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test orderDataMutableUpdater logic 1`] = `
Object {
  "idOrderMap": Map {
    "b11b8a69" => Object {
      "customer": "Karina Lopez",
      "destination": "89548 Andrew Road Apt. 797, Lake Phillipville, CA 92007",
      "event_name": "COOKED",
      "id": "b11b8a69",
      "item": "Fish tacos",
      "price": 8700,
      "sent_at_second": 30,
    },
    "0b63508d" => Object {
      "customer": "Heather Hughes",
      "destination": "7465 Willis Fork, Lake Gabrielle, CA 92816",
      "event_name": "CREATED",
      "id": "0b63508d",
      "item": "Berry smoothie",
      "price": 8029,
      "sent_at_second": 5,
    },
  },
  "priceIdMap": Map {
    8700 => Set {
      "b11b8a69",
    },
    8029 => Set {
      "0b63508d",
    },
  },
}
`;

exports[`test orderDataMutableUpdater logic 2`] = `
Object {
  "idOrderMap": Map {
    "b11b8a69" => Object {
      "customer": "Karina Lopez",
      "destination": "89548 Andrew Road Apt. 797, Lake Phillipville, CA 92007",
      "event_name": "DRIVER_RECEIVED",
      "id": "b11b8a69",
      "item": "Fish tacos",
      "price": 8700,
      "sent_at_second": 30,
    },
    "0b63508d" => Object {
      "customer": "Heather Hughes",
      "destination": "7465 Willis Fork, Lake Gabrielle, CA 92816",
      "event_name": "CREATED",
      "id": "0b63508d",
      "item": "Berry smoothie",
      "price": 8029,
      "sent_at_second": 5,
    },
  },
  "priceIdMap": Map {
    8700 => Set {
      "b11b8a69",
    },
    8029 => Set {
      "0b63508d",
    },
  },
}
`;

exports[`test orderDataMutableUpdater logic 3`] = `
Object {
  "idOrderMap": Map {
    "b11b8a69" => Object {
      "customer": "Karina Lopez",
      "destination": "89548 Andrew Road Apt. 797, Lake Phillipville, CA 92007",
      "event_name": "DRIVER_RECEIVED",
      "id": "b11b8a69",
      "item": "Fish tacos",
      "price": 8700,
      "sent_at_second": 30,
    },
    "0b63508d" => Object {
      "customer": "Heather Hughes",
      "destination": "7465 Willis Fork, Lake Gabrielle, CA 92816",
      "event_name": "CREATED",
      "id": "0b63508d",
      "item": "Berry smoothie",
      "price": 8029,
      "sent_at_second": 5,
    },
    "b11b8a69xxx" => Object {
      "customer": "Karina Lopez",
      "destination": "89548 Andrew Road Apt. 797, Lake Phillipville, CA 92007",
      "event_name": "CREATED",
      "id": "b11b8a69xxx",
      "item": "Fish tacos",
      "price": 8700,
      "sent_at_second": 30,
    },
  },
  "priceIdMap": Map {
    8700 => Set {
      "b11b8a69",
      "b11b8a69xxx",
    },
    8029 => Set {
      "0b63508d",
    },
  },
}
`;
